### PR TITLE
Performance Test Proposal

### LINKED ISSUE #182: Evaluate KubeEdge scalability
KubeEdge allows admin to manage large scale of edge nodes from cloud. With Edgecontroller, cloudhub/edgehub and metadata store at edge side, we are extending the potential capacity. 

To prove the advantages, this is to book an item for us evaluating the 
- latency & throughput for cloudhub/edgehub
- scalability of kubeEdge cloud side

As a result, we'll also evaluate the future improvement items.

### PR DESCRIPTION
**Pull request by edisonxiang**
What type of PR is this?
/kind design

What this PR does / why we need it:
Proposal for performance test scenarios and test cases.

Which issue(s) this PR fixes:
Partially Fixes #182

Special notes for your reviewer:
More test scenarios and test cases are welcome

### CODE REVIEW

**Comment by sids-b** [comment]
cc @qizha @Baoqiang-Zhang @CindyXing @m1093782566 @kevin-wangzefeng

File: docs/proposals/perf.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+  Edge Nodes numbers are one of `[1, 10, 20, 50, 100, 200]`.
+
+  Pods numbers per Edge Node are one of `[1, 2, 5, 10, 20]`.
+
+As the result of the pod startup time, we can evaluate scalability of KubeEdge Edge Nodes.
+Measure the capacity of Edge Nodes can be supported by KubeEdge Cloud Part.
+<img src="../images/perf/perf-multi-edgenodes.png">
+
+The simulation of generating virtual KubeEdge Edge Nodes are being analyzed with `Kubemark` framework.
```
Comment (rohitsardesai83) [inline docs/proposals/perf.md]: we need to have a design on how are we planning to simulate edge node . will it be covered in a separate PR?

File: docs/proposals/perf.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  Different numbers of Edge Nodes need be tested.
+
+  Edge Nodes numbers are one of `[1, 10, 20, 50, 100, 200]`.
+
+  This test case ends with all Edge Nodes are in `Ready` status.
+
+### 2. Create Device/Device Model from Cloud
+<img src="../images/perf/perf-create-device.png">
+
+This scenario is expected to measure the northbound API of KubeEdge.
```
Comment (rohitsardesai83) [inline docs/proposals/perf.md]: since device model and devices are CRD's which are installed via APIs exposed by the API server , I am not sure whether we should measure these .  when we create a device referring to a device model , the CRD APIs exposed by the K8S API server will be invoked. We don't implement these APIs but we define what is the API .  We need to think through more on what exactly we need to measure from cloud side ? 
Yes , when the device controller watches devices created in cloud and creates config maps for the devices and pushes them to the edge , this end to end latency could be measured.

File: docs/proposals/perf.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+### Goals
+
+* Benchmark the performance against the following Service Level Objectives:
+  * Latency: time cost from the moment when the server gets the request to last byte of response sent to the users.
+  * Throughput: measure how many requests can be served within given time.
+  * Scalability: potential scaling capacity under different load conditions.
+  * CPU Usage: measure the cpu usage of KubeEdge under different load conditions.
+  * Memory Usage: measure the memory usage of KubeEdge under different load conditions.
+* Performance test can be able to run against both dockerized and un-dockerized version of KubeEdge.
```
Comment (rohitsardesai83) [inline docs/proposals/perf.md]: can be able to run  -> should be able to run

File: docs/proposals/perf.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+### Non-goals
+
+* To design the specific implementation of single performance test.
+
+## Proposal
+
+### Test Framework
```
Comment (rohitsardesai83) [inline docs/proposals/perf.md]: We need to have a detailed design for the framework .

File: docs/proposals/perf.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+<img src="../images/perf/perf-test-framework.png">
+
+### Test Tools
```
Comment (rohitsardesai83) [inline docs/proposals/perf.md]: are we planning to re-use these frameworks or take reference and build our own ?

File: docs/proposals/perf.md (hunk 2 above)
Comment (m1093782566) [reply docs/proposals/perf.md]: +1

File: docs/proposals/perf.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+* [Fortio](https://github.com/fortio/fortio)
+
+### Metrics Tools
+* [Prometheus](https://github.com/prometheus/prometheus)
+* [Grafana](https://github.com/grafana/grafana)
+
+## Test Scenarios
+
+### 1. Edge Nodes join in K8S Cluster
+<img src="../images/perf/perf-edgenodes-join-cluster.png">
```
Comment (m1093782566) [inline docs/proposals/perf.md]: I think test client will watch the node status instead of periodically `get`.

File: docs/proposals/perf.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+### 3. Report Device Status to Edge
+<img src="../images/perf/perf-report-devicestatus.png">
+
+This scenario is expected to measure the southbound API of KubeEdge.
+
+Test Cases:
+* Measure the latency between KubeEdge Edge Part and device.
+
+  Device numbers are one of `[1, 10, 20, 50, 100, 200]`.
```
Comment (m1093782566) [inline docs/proposals/perf.md]: Do you mean register devices to the same node?

File: docs/proposals/perf.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+* Measure the throughput between KubeEdge Edge Part and device.
+
+  Device numbers are one of `[1, 10, 20, 50, 100, 200]`.
```
Comment (m1093782566) [inline docs/proposals/perf.md]: Do you mean register devices to the same node at the same time?

File: docs/proposals/perf.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+This scenario is expected to measure the performance of KubeEdge from Cloud to Edge.
+
+Test Cases:
+* Measure the pod startup time with docker images in the Edge Node.
+
+  Edge Nodes numbers are one of `[1, 10, 20, 50, 100, 200]`.
+
+  Pods numbers per Edge Node are one of `[1, 2, 5, 10, 20]`.
+
+* Measure the pod startup time without docker images in the Edge Node.
```
Comment (m1093782566) [inline docs/proposals/perf.md]: Probably we should not include image download latency as it's out of this scope.

File: docs/proposals/perf.md [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+This scenario is expected to measure the e2e performance of KubeEdge.
+
+Test Cases:
+* Measure e2e latency.
+
+  Edge Nodes numbers are one of `[1, 10, 20, 50, 100, 200]`.
+
+  Device numbers are one of `[1, 10, 20, 50, 100, 200]`.
+
+* Measure CPU and Memory Usage of KubeEdge Cloud Part.
```
Comment (m1093782566) [inline docs/proposals/perf.md]: I think CPU and memory usage should be measured in some of the above scenarios.

File: docs/proposals/perf.md [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+<img src="../images/perf/perf-multi-edgenodes.png">
+
+The simulation of generating virtual KubeEdge Edge Nodes are being analyzed with `Kubemark` framework.
+
+The following deployment types of KubeEdge Edge Nodes can be used for test:
+* Deploy virtual KubeEdge Edge Nodes in single VM.
+* Deploy actual KubeEdge Edge Node per dedicated VM.
+
+### 5. Update Device Twin State from Cloud to Device
+<img src="../images/perf/perf-update-devicetwin.png">
```
Comment (m1093782566) [inline docs/proposals/perf.md]: Watch the result instead of periodically get.

File: docs/proposals/perf.md (hunk 1 above)
Comment (pavan187) [reply docs/proposals/perf.md]: @edisonxiang , can you please push the different deployment scenario's added in design doc, where i have explained the edge node simulations under the deployments.

Comment (CindyXing) [reply docs/proposals/perf.md]: Echo what @rohitsardesai83  called out here. We are using kubemark to test control plane. Need to evaluate whether kubemark's node simulation works or not. 

Besides, we need to evaluate the edge node local performance. After edgehub receives metadata,  how long it takes for edged to take effect. Also how many pods can we run on edge node provided different level constraint of resources (1CPU, 256MB； 2CPU 1GB, etc.)

File: docs/proposals/perf.md (hunk 4 above)
Comment (CindyXing) [reply docs/proposals/perf.md]: +1

**Comment by CindyXing** [comment]
From scalability/performance-wise, it is critical to evaluate edgehub/cloudhub+controller communication. For this, if we are to use kubemark, some update would be needed. 

Some of the scenarios 
E2E performance
1. Many deployments were made at master
2. A node is connected
3. Evaluate how long for edge node to obtain all the messages; and whether it is ok from customer's perspective
4. Edge node crashes 
5. More deployments were made at master node 
6. Edge node is re-connected
7. Evaluate how long edge node to obtain the messages

Scalability
1. How many edge nodes can concurrently connects to master through websocket. And cloudhub sending/receiving messages. Measure in the sense of latency, throughput, cpu and memory usage

2. How many pods and devices can be handled at edge 

The performance and scalability of cloud K8s can be used as the base to compare. Please watch out whether we could support more edge nodes than cloud k8s. 

This effort is critical for us to evaluate the benefits of using KubeEdge.

File: docs/proposals/perf.md (hunk 1 above)
Comment (m1093782566) [reply docs/proposals/perf.md]: +1

File: docs/proposals/perf.md (hunk 7 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: yup

File: docs/proposals/perf.md (hunk 8 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: yup

File: docs/proposals/perf.md (hunk 9 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: OK. Thanks

File: docs/proposals/perf.md (hunk 1 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: Thanks, we will update later

File: docs/proposals/perf.md (hunk 2 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: we will look over this scenario

File: docs/proposals/perf.md (hunk 4 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: Thanks, we will update later

File: docs/proposals/perf.md (hunk 5 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: We will have detailed design

File: docs/proposals/perf.md (hunk 3 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: Got it

**Comment by sids-b** [comment]
> From scalability/performance-wise, it is critical to evaluate edgehub/cloudhub+controller communication. For this, if we are to use kubemark, some update would be needed.
> 
> Some of the scenarios
> E2E performance
> 
> 1. Many deployments were made at master
> 2. A node is connected
> 3. Evaluate how long for edge node to obtain all the messages; and whether it is ok from customer's perspective
> 4. Edge node crashes
> 5. More deployments were made at master node
> 6. Edge node is re-connected
> 7. Evaluate how long edge node to obtain the messages
> 
> Scalability
> 
> 1. How many edge nodes can concurrently connects to master through websocket. And cloudhub sending/receiving messages. Measure in the sense of latency, throughput, cpu and memory usage
> 2. How many pods and devices can be handled at edge
> 
> The performance and scalability of cloud K8s can be used as the base to compare. Please watch out whether we could support more edge nodes than cloud k8s.
> 
> This effort is critical for us to evaluate the benefits of using KubeEdge.

+1

**Comment by sids-b** [comment]
ping @edisonxiang @pavan187 @samy2019 for updates.

**Comment by kevin-wangzefeng** [comment]
Let me know when it's ready for another round review.

**Comment by m1093782566** [comment]
@edisonxiang Please fix the outstanding comments.

File: docs/proposals/perf.md (hunk 1 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: Thanks @CindyXing. We have done the research for kubemark.
We found it is not suitable for us to do perf test in kubeedge.
We will not use kubemark. 
Please find more from this link.
https://docs.google.com/presentation/d/1oVB92dLNKMMVJOzfvsdrRthpPwAJxSDlrwe0dkG_R4o/edit#slide=id.g563407762e_0_0

**Comment by edisonxiang** [comment]
@qizha @Baoqiang-Zhang @CindyXing @m1093782566 @kevin-wangzefeng @rohitsardesai83 @pavan187 
The proposal is updated. PTAL. Thanks.

File: docs/proposals/perf.md (hunk 1 above)
Comment (CindyXing) [reply docs/proposals/perf.md]: That makes sense. Thanks @edisonxiang. One thing, can we run the hollowedgenode at a separate vm? This way we could launch more edge nodes and not impact the k8s master. I might misunderstand the slide 3 chart.

File: docs/proposals/perf.md [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+| Subject                        | Description                                  |
+|--------------------------------|----------------------------------------------|
+| OS                             |  Ubuntu 18.04 server 64bit                   |
+| Docker Version                 |  v17.09                                      |
+| CPU                            |  32vCPUs                                     |
+| RAM                            |  128GB                                       |
+| Disk Size                      |  40GB                                        |
+| Count                          |  1...N                                       |
+
+These VMs are used to deploy numbers of KubeEdge Edge Nodes pods which are running Edged and EdgeHub and so on.
```
Comment (rohitsardesai83) [inline docs/proposals/perf.md]: will we have 1 vm per edge node pod ? I thought the concept of hollow node was to avoid have a large number of vm's as we are just running hollow edge node pods which are mimicking a real vm .

Comment (edisonxiang) [reply docs/proposals/perf.md]: @rohitsardesai83, no, we will have lots of edge node pods in 1 vm as the deployment graph.
But we need to consider the vcpu and ram of the vm. because 1 vm is not enough to run all of the edge node pods. We will clarify it by some notes

Comment (pavan187) [reply docs/proposals/perf.md]: @rohitsardesai83 , for KubeEdge perofrmance setup is similar to kubemark setup, where they simulate no.of hollow-nodes in pods and deploy on k8s cluster. in kubeedge also we do the similar kind of simulation for creating kubeedge node in pod and deploy the pods through deploment object.

Comment (edisonxiang) [reply docs/proposals/perf.md]: @rohitsardesai83 more details are added in the lasest commit with @pavan187

**Comment by edisonxiang** [comment]
@CindyXing Looking forward your reply and suggestions. Thank you!

**Comment by m1093782566** [comment]
Apart from hollow-node, is it possible to mock devices, i.e. bluetooh device, opc-ua device and modbus device? 

@pavan187 @edisonxiang thoughts?

File: docs/proposals/perf.md [hunk 13]
```diff
@@ -103,9 +108,73 @@ This VM is used to run KubeEdge Cloud Part Services including Edge Controller an
...
+KubeEdge Performance Test Framework will be designed based on the **Gomega** and **Ginkgo**.
+
+The Performance Test Framework mainly relates to Utils Library and different types of tests:
+- E2E Test
+- Latency Test
+- Load Test
+- Scalability Test
+- ...
+
+E2E Test Sample:
```
Comment (CindyXing) [inline docs/proposals/perf.md]: This is a great test case. Can you help me understand how you measure the time duration for latency and throughput?

**Comment by edisonxiang** [comment]
> Apart from hollow-node, is it possible to mock devices, i.e. bluetooh device, opc-ua device and modbus device?
> 
> @pavan187 @edisonxiang thoughts?

@m1093782566 Yeah, actually in the `scenario 3: report-device-status-to-edge` and `scenario 5: Update Device Twin State from Cloud to Device`, we will try to measure how many devices can be supported by the edge nodes. We will try to mock different devices, firstly we can mock bluetooth device as we have bluetooth device mapper. This work will be done in the detail test case:). Thanks.

**Comment by edisonxiang** [comment]
> For example, something like https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md
> 
> Ideally the perf/scalability effort can help us identifying critical design or code issues.

Thanks @CindyXing. We will take a look at this point.

**Comment by m1093782566** [comment]
> @m1093782566 Yeah, actually in the scenario 3: report-device-status-to-edge and scenario 5: Update Device Twin State from Cloud to Device, we will try to measure how many devices can be supported by the edge nodes. We will try to mock different devices, firstly we can mock bluetooth device as we have bluetooth device mapper. This work will be done in the detail test case:). Thanks.

cc @lidiyag for mocking devices in katacoda tutorials.

File: docs/proposals/perf.md (hunk 13 above)
Comment (pavan187) [reply docs/proposals/perf.md]: @CindyXing , Above is sample test case given deploy application to kubeeedge node and check the pods are coming into ready state, above sample case doesnt have measurement criteria for pod startup time, that logic of measuring the latency yet be added.

while doing measurement from framework, we will take the pod created time(t1) stamp from test cases, and get the timestamp when pod gets into the running state(t2), from the t1 and t2, we will calculate the time taken for pod to come up and Running.

Coming to throughput, From the tests we do the parallel depployments using go-routunes (Ginkgo has build-in support to go routines) on single-node and calculate the CPU,Memory consupmtion for diffrent no.of deployments done.

**Comment by edisonxiang** [comment]
@CindyXing @m1093782566 @pavan187 
The thresholds are added, PTAL, Thanks very much.

File: docs/proposals/perf.md (hunk 13 above)
Comment (CindyXing) [reply docs/proposals/perf.md]: What if I want the specific latency for cloudhub/edgehub message delivery? Measuring e2e is a blackbox validation of the perf. Here we should do whitebox. That's how we could quickly identify any potential issue(s). Thanks

**Comment by CindyXing** [comment]
Thanks @edisonxiang! Overall it looks pretty good now. 
The only pending question I have is how we measure the latency/throughput between cloudhub/edgehub. As mentioned many times in my comments, it is the most critical part for KubeEdge scalability/performance. We should have a whitebox measurement against.

File: docs/proposals/perf.md (hunk 13 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: Thanks @CindyXing. This scenario is added in the latest commit.

**Comment by edisonxiang** [comment]
Thanks @CindyXing. This scenario is added in the latest commit.
Please take a look at the scenario 6. Add Pod from CloudHub to EdgeHub.

File: docs/proposals/perf.md [hunk 14]
```diff
@@ -298,6 +298,29 @@ Test Cases:
...
+Actually this is not an E2E Test scenario for KubeEdge,
+but the message delivery channel between CloudHub to EdgeHub may be our bottleneck.
+Currently we are using web socket as the communication protocal between Cloud and Edge.
+In the following test cases, we need to mock the behaviors of CloudHub and EdgeHub,
+and the simulation messages of adding pod will be sent to EdgeHub,
+and the simulation messages of pod status will be sent back to CloudHub.
+so that we can get the exact latency and throughput between CloudHub and EdgeHub.
+
+Test Cases:
+* Measure the latency between KubeEdge CloudHub and KubeEdge EdgeHub.
```
Comment (CindyXing) [inline docs/proposals/perf.md]: Please pay good attention on
-  when bunch of edge nodes got disconnected & then reconnect. 
-  cloudhub CPU/memory when there are large scale of edge nodes where cloudhub is sending/receiving messages.

Comment (edisonxiang) [reply docs/proposals/perf.md]: Thanks @CindyXing . The CPU/memory is added in this scenario.
we will consider disconnected & then reconnect case once we have the original test data for this scenario.

**Comment by edisonxiang** [comment]
Thanks everyone. 
PR is squashed.
@m1093782566 PTAL.

**Comment by m1093782566** [comment]
Defer to @CindyXing for merging.

**Comment by edisonxiang** [comment]
@CindyXing PTAL. Thank you:)

File: docs/proposals/perf.md [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+| Docker Version                 |  v17.09                                      |
+| CPU                            |  32vCPUs                                     |
+| RAM                            |  128GB                                       |
+| Disk Size                      |  40GB                                        |
+| Count                          |  1...N                                       |
+
+These VMs are used to deploy numbers of KubeEdge Edge Nodes pods which are running Edged and EdgeHub and so on.
+We will adjust the Count of VMs based on the KubeEdge Edge Nodes numbers.
+
+KubeEdge Performance Test setup is similar with K8S KubeMark setup,
```
Comment (kevin-wangzefeng) [inline docs/proposals/perf.md]: The KubeMark like deployment is for simulating the control-plane scalability.
You may also describe the simple setup that used for testing edgenode CPU/Mem usage etc.

File: docs/proposals/perf.md [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+where they simulate numbers of hollow-nodes in pods and deploy on K8S Cluster.
+In KubeEdge we also do the similar kind of simulation for creating KubeEdge Edge Nodes
+in pods and deploy the pods through deployment. Our pod takes up resources as below:
```
Comment (kevin-wangzefeng) [inline docs/proposals/perf.md]: You may call the kubeEdge nodes running as pods an "edge-node-pod", and a pod running applications an "application-pod" to avoid confusion.

File: docs/proposals/perf.md [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+- 1 pod : 0.10 vCPU & 250MB RAM
+
+With KubeEdge pod deployment we can accomodate 10 pods/1vCPU approximately.
+Base on the above K8S Nodes flavor, the CPU and RAM are 32vCPU and 128GB respectively.
+Per K8S Node we should be able to deploy 320 pods(KubeEdge Edge Nodes)/32vCPU
+and RAM consumption would be around 80GB. If we have 5 K8S Nodes with the similar flavor,
+on a whole we should be able to deploy 1500 pods(KubeEdge Edge Nodes)/5 K8S Nodes.
```
Comment (kevin-wangzefeng) [inline docs/proposals/perf.md]: 1. Will the applications' containers running inside the edge-node-pods or co-located with it on the host node? -- Do you plan to use docker in docker for edge nodes or just mount docker.sock into the edge-node container?
2. A K8s node should not be filled up with edge-node-pods, which would result in no application-pods able to be served -- the 320 pods /32vCPU here should be in sum of edge-node-pods + application-pods that running on the edge-nodes.

File: docs/proposals/perf.md [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+
```
Comment (kevin-wangzefeng) [inline docs/proposals/perf.md]: Any resource requirements of the test client side? To avoid client side throughput limitation etc.

**Comment by edisonxiang** [comment]
@CindyXing Thanks. I will

File: docs/proposals/perf.md [hunk 19]
```diff
@@ -0,0 +1,364 @@
...
+## Proposal
+
+### Performance Test Deployment
+<img src="../images/perf/perf-deploy-type.png">
+
+Every running KubeEdge Performance Test setup looks like the following:
+
+1. A real K8S Cluster that has master and nodes.
+2. A separated VM where the KubeEdge Cloud Part Services are running, including Edge Controller and Cloud Hub and so on.
+3. Test Client(E2E) will build KubeEdge EdgeNode image and put into the Docker Hub repository.
```
[the remark this answers is not in the document]
Comment (pavan187) [reply docs/proposals/perf.md]: @kevin-wangzefeng , OK...will keep the pre-requisite to the user to keep the image ready and keep it in a reachable container image registry.

File: docs/proposals/perf.md [hunk 20]
```diff
@@ -0,0 +1,364 @@
...
+    * [Performance Test Framework](#performance-test-framework)
+    * [Performance Test Metrics Tools](#performance-test-metrics-tools)
+    * [Performance Test Scenarios](#performance-test-scenarios)
+  * [Thresholds](#thresholds)
+
+## Motivation
+
+Currently KubeEdge test is focused on automated test suites for unit, integration and E2E test and validation.
+KubeEdge allows the users to manage large scale of edge nodes, devices from cloud.
+A set of test specifically performance test can be used to determine the non-functional
```
[the remark this answers is not in the document]
Comment (edisonxiang) [reply docs/proposals/perf.md]: Done. Thanks

File: docs/proposals/perf.md [hunk 21]
```diff
@@ -0,0 +1,364 @@
...
+
+### Goals
+
+* Benchmark the performance against the following Service Level Objectives:
+  * Latency: time cost from the moment when the server gets the request to last byte of response sent to the users.
+  * Throughput: measure how many requests can be served within given time.
+  * Scalability: potential scaling capacity under different load conditions.
+  * CPU Usage: measure the cpu usage of KubeEdge under different load conditions.
+  * Memory Usage: measure the memory usage of KubeEdge under different load conditions.
+* Performance test should be able to run against both dockerized and un-dockerized version of KubeEdge.
```
[the remark this answers is not in the document]
Comment (edisonxiang) [reply docs/proposals/perf.md]: Done. Thanks

File: docs/proposals/perf.md [hunk 22]
```diff
@@ -0,0 +1,364 @@
...
+
+### Non-goals
+
+* To design the specific implementation of single performance test.
```
[the remark this answers is not in the document]
Comment (edisonxiang) [reply docs/proposals/perf.md]: Done. Thanks

File: docs/proposals/perf.md [hunk 23]
```diff
@@ -0,0 +1,364 @@
...
+As a result, we can also evaluate the future improvement items for KubeEdge.
+
+This proposal lists the possible performance test scenarios and test cases for KubeEdge.
+
+### Goals
+
+* Benchmark the performance against the following Service Level Objectives:
+  * Latency: time cost from the moment when the server gets the request to last byte of response sent to the users.
+  * Throughput: measure how many requests can be served within given time.
+  * Scalability: potential scaling capacity under different load conditions.
```
[the remark this answers is not in the document]
Comment (edisonxiang) [reply docs/proposals/perf.md]: Done. Thanks

File: docs/proposals/perf.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (edisonxiang) [reply docs/proposals/perf.md]: Done. Thanks

File: docs/proposals/perf.md (hunk 15 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: The deployment figure is updated. Thanks

File: docs/proposals/perf.md (hunk 17 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: Thanks. Very important catch for us.
We will use docker in docker for deployment.

File: docs/proposals/perf.md (hunk 18 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: Added it. Thanks

File: docs/proposals/perf.md (hunk 16 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: Thanks. The sentence is updated

File: docs/proposals/perf.md [hunk 24]
```diff
@@ -0,0 +1,364 @@
...
+* Performance test should be able to run against both dockerized and un-dockerized version of KubeEdge.
+
+### Non-goals
+
+* To design the specific implementation of single performance test.
+
+## Proposal
+
+### Performance Test Deployment
+<img src="../images/perf/perf-deploy-type.png">
```
[the remark this answers is not in the document]
Comment (edisonxiang) [reply docs/proposals/perf.md]: Thanks. The deployment figure is updated.

**Comment by edisonxiang** [comment]
@kevin-wangzefeng 
Thanks your comments.
The deployment figure is updated.
PTAL

File: docs/proposals/perf.md [hunk 25]
```diff
@@ -0,0 +1,381 @@
...
+
+#### 1. Edge Nodes join in K8S Cluster
+<img src="../images/perf/perf-edgenodes-join-cluster.png">
+
+Test Cases:
+* Measure Edge Nodes join in K8S Cluster startup time.
+
+  Different numbers of Edge Nodes need be tested.
+
+  Edge Nodes numbers are one of `[1, 10, 20, 50, 100, 200...]`.
```
Comment (rohitsardesai50) [inline docs/proposals/perf.md]: if the numbers are same for all the test cases , it is better to mention only once before describing the tests.

File: docs/proposals/perf.md [hunk 26]
```diff
@@ -0,0 +1,381 @@
...
+
+#### 3. Report Device Status to Edge
+<img src="../images/perf/perf-report-devicestatus.png">
+
+This scenario is expected to measure the southbound API of KubeEdge.
+
+Test Cases:
+* Measure the latency between KubeEdge Edge Part and device.
+
+  Device numbers per Edge Node are one of `[1, 10, 20, 50, 100, 200...]`.
```
Comment (rohitsardesai50) [inline docs/proposals/perf.md]: see previous comment , avoid repetitiveness. thanks. applies to all such occurrences

File: docs/proposals/perf.md [hunk 27]
```diff
@@ -0,0 +1,381 @@
...
+E.g. Bluetooth, MQTT, ZigBee, BACnet and Modbus and so on.
+Currenly less than 20ms latency can be accepted in Edge IoT scenario.
+Two kinds of test cases can be adopted: emulators of different devices and actual devices.
+
+#### 4. Application Deployment from Cloud to Edge
+<img src="../images/perf/perf-app-deploy.png">
+
+This scenario is expected to measure the performance of KubeEdge from Cloud to Edge.
+The docker image download latency is not included in this scenario.
+In the following test cases, we need to make sure that docker images have already been in the Edge Node.
```
Comment (rohitsardesai50) [inline docs/proposals/perf.md]: have already been in the Edge Node -> are already downloaded on the edge node.

File: docs/proposals/perf.md (hunk 25 above)
Comment (rohitsardesai83) [inline docs/proposals/perf.md]: if the numbers are same for all the test cases , it is better to mention only once before describing the tests.

File: docs/proposals/perf.md (hunk 26 above)
Comment (rohitsardesai83) [inline docs/proposals/perf.md]: see previous comment . avoid repetitive content

File: docs/proposals/perf.md [hunk 28]
```diff
@@ -0,0 +1,381 @@
...
+
+As [this K8S document](https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md) mentions,
+since 1.6 release K8S can support 5000 Nodes and 150000 Pods in single cluster.
+KubeEdge is based on K8S Master, the difference between K8S Nodes and KubeEdge Edge Nodes
+is that KubeEdge Edge Nodes are not directly connected with the K8S Master like the K8S Nodes do.
+The KubeEdge Cloud Part connects K8S Master with KubeEdge Edge Nodes.
+Certainly the KubeEdge Edge Nodes are light weight and making use of less resources like CPU, Memory.
+
+Currently about KubeEdge we have no performance data which can make comparison with the other systems.
+But we can measure the performance and scalability for KubeEdge by the Performance Test data.
```
Comment (rohitsardesai83) [inline docs/proposals/perf.md]: by the Performance Test data -> using the Performance Test data.

File: docs/proposals/perf.md (hunk 27 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: Thanks. will update later

File: docs/proposals/perf.md (hunk 28 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: Thanks. will update later

File: docs/proposals/perf.md (hunk 25 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: I will Thanks

File: docs/proposals/perf.md (hunk 26 above)
Comment (edisonxiang) [reply docs/proposals/perf.md]: I will Thanks

**Comment by rohitsardesai83** [comment]
lgtm. thanks @edisonxiang , @pavan187 for your efforts !
